### DISCUSSION TITLE
Send a file as request response and download it on the client

### DISCUSSION THREAD

**Discussion by vitorbertolucci** [vercel/next.js#16858] [Help]
I'm trying to download a zip file from an API route on the client but the file I get is corrupted.

I've checked both https://github.com/vercel/next.js/discussions/16671 and https://github.com/vercel/next.js/discussions/15453

Here is my backend code

```
export default async (req, res) => {
  ...
  const filePath = path.resolve(process.cwd(), "file.zip");
  const stat = fs.statSync(filePath);

  res.writeHead(200, {
    "Content-Type": "application/zip",
    "Content-Length": stat.size
  });

  const readStream = fs.createReadStream(filePath);
  readStream.pipe(res);
  ...
};

export const config = {
  api: {
    bodyParser: false
  }
};

```

And here is my client code

```
const handleUploadFiles = async () => {
    const formData = new FormData();
    formData.append("file1", file);

    const data = await api.post("/api/shapefile", formData);
    const blob = new Blob([data.data], { type: "application/zip" });
    saveAs(blob, "teste.zip");
};
```

the saveAs method comes from file-saver package. The zip file is downloaded, but it won't open.

In case anyone asks why I'm trying to do it, here it goes: The user will upload a .kml file, this file will be sent to the API, converted into shapefile and then zipped. Finally, the zipped file will be sent to the user so he can save it on his device.

I'm not sure if I'm doing it the best way, so I'm entierly opened to suggestions.

Thanks in advance.

**Comment by Noitidart** [comment]
Were you able to figure out if this is best way?
